templates/java/org/yarp/Loader.java.erb

```diff
@@ -30,15 +30,26 @@ public class Loader {
         byte[] get(ByteBuffer buffer, int oneBasedIndex) {
             int index = oneBasedIndex - 1;
             byte[] constant = cache[index];
+
             if (constant == null) {
                 int offset = bufferOffset + index * 8;
                 int start = buffer.getInt(offset);
                 int length = buffer.getInt(offset + 4);
 
                 constant = new byte[length];
-                System.arraycopy(source, start, constant, 0, length);
+
+                if ((start & (1 << 31)) == 0) {
+                    System.arraycopy(source, start, constant, 0, length);
+                } else {
+                    int position = buffer.position();
+                    buffer.position(start - (1 << 31));
```

Comment:
```suggestion
                    buffer.position(start & 0x7fffffff);
```
since `1 << 31` doesn't fit in a signed 32-bit int so it feels rather hacky (also because `start` is never >= 1<<31).

Reply:
Updated to use `Math.abs(start)`

Reply:
I don't think that's correct:
```
jshell> int start = -3
start ==> -3

jshell> start & 0x7fffffff
$2 ==> 2147483645

jshell> Math.abs(start)
$3 ==> 3
```

Reply:
We should probably improve the Java toString() output, then the `build-java` CI job wouldn't catch it but it would be easy to check if it looks correct or not.
Alternatively when TruffleRuby starts using this we should notice as well in the daily CI job.

Reply:
The only code that seems to trigger `if (constant->owned) {` seems ParseSerializeTest.
CallNode and operator nodes seem to use type `string` and not `constant` yet.

Reply:
+ that case doesn't actually use `foo` and adding `a = foo if rand < 0` at the top of the file considers foo as a call but it should be a LocalVariableReadNode. So this feels like it needs more testing and that's why this bug does not really happen yet.

Reply:
It'll be fixed by https://github.com/ruby/yarp/pull/1466

Reply:
Nice, I'm happy the better toString caught that :)

Reply:
Ah actually just deserializing caught it, since https://github.com/ruby/yarp/pull/1454 isn't merged yet.